### PR TITLE
Update the behavior of the cached undo/redo stack

### PR DESCRIPTION
**Pull request by youknowriad**
## What?

Until now, any change to what we call "transient properties" (like blocks, selection) was considered as a change that doesn't create an undo/redo step. This PR makes the assumption that this heuristic is wrong. IMO any change to "blocks" for instance should actually create an undo/redo step.

That said, we do have some smart behavior in places where we want sometimes to cache changes and only create an undo step after some time, for instance when typing quickly in a RichText component. For these cases, the `isCached` flag is the solution proposed by this PR.

Also, this PR enables a new behavior to unblock #51201 

Basically, with the new flag, it's possible to have calls to `editEntityRecords` updating non-transient properties like "meta" or "title" and cache these changes and only create an undo/redo once for these changes potential consecutive changes. 

## Testing Instructions

1- Open the post editor
2- Try doing a few different things
3- Make sure that undo/redo still behaves similarly to trunk.

(We also have e2e tests to validate that)

### CODE REVIEW

**Comment by Mamaduka** [comment]
Will this also unblock #37171?

**Comment by youknowriad** [comment]
@Mamaduka It will not solve it but will provide us a way to do it yes.

File: docs/explanations/architecture/entities.md [hunk 1]
```diff
@@ -56,8 +56,14 @@ For example, let's say a user edits the title of a post, followed by a modificat
...
 
-So by default, `core-data` store considers all modifications to properties that are marked as "transient" (like the `blocks` property in the post entity) as transient modifications. It keeps these modifications outside the undo/redo stack in what is called a "cache" of modifications and these modifications are only stored in the undo/redo stack when we explicitely call `__unstableCreateUndoLevel` or when the next non-transient modification is performed.
+Cached changes are kept outside the undo/redo stack in what is called a "cache" of modifications and these modifications are only stored in the undo/redo stack when we explicitely call `__unstableCreateUndoLevel` or when the next modification is not a cached one.
+
+By default all calls to `editEntityRecord` are considered "non-cached" unless the `isCached` option is passed as true. Example:
+
+```js
+wp.data.dispatch( 'core' ).editEntityRecord( 'postType', 'post', 1, { title: 'Hello World' }, { isCached: true } );
```
Comment (paulopmt1) [inline docs/explanations/architecture/entities.md:68]: Thanks for that example!

**Review by paulopmt1 — APPROVED** [review]
Tested blocks, images, templates, and text edition and everything worked as expected.

> it's possible to have calls to editEntityRecords updating non-transient properties like "meta" or "title"

That's great to hear! I'm planning [to tackle this one](https://github.com/WordPress/gutenberg/issues/37171) in my next 20% time to see if I really understand undo/redo now 😅

File: packages/core-data/src/entity-provider.js [hunk 2]
```diff
@@ -275,7 +275,7 @@ export function useEntityBlockEditor( kind, name, { id: _id } = {} ) {
 			const edits = { blocks: newBlocks, selection };
 			registry.batch( () => {
 				updateFootnotes( edits.blocks );
-				editEntityRecord( kind, name, id, edits );
+				editEntityRecord( kind, name, id, edits, { isCached: false } );
```
Comment (ellatrix) [inline packages/core-data/src/entity-provider.js]: @youknowriad I rebase this PR to try it with footnotes, but undo still doesn't seem batched. Do we need to update the `editEntityRecord` call in `useEntityProp` as well?

Comment (youknowriad) [reply packages/core-data/src/entity-provider.js]: That change is not correct. For this PR to only create one undo level for footnotes, you need one of the `editEntityRecord` calls (either footnotes one or actual change) to be `isChached: true` (isCached: false is the default)

Comment (ellatrix) [reply packages/core-data/src/entity-provider.js]: Oh, sorry. I just rebased with the previous change.

**Review by ellatrix — COMMENTED** [review]
If you reorder two paragraph blocks with a footnote in each, undo will now revert the change as a whole correctly, but only after you undo, redo, and undo again. Somehow on the first time it messes up. 🤔 

If a page already has footnotes, and you add another one, undoing the addition right away will only remove the footnote from the block. But somehow when you press redo, then undo again, it will undo the entire change. So same problem happening here.

But if you start off with no footnotes, and a footnotes block needs to be inserted, the first undo seems to do nothing, and consecutive redo/undo does it in two steps. Not sure why, since everything is batched:

https://github.com/WordPress/gutenberg/blob/1be8fa3c5f63a0c9a096478805665724417512aa/packages/block-library/src/footnotes/format.js#L46-L100

**Comment by youknowriad** [comment]
I'm happy to solve the footnotes undo as well, but just wanted to note that this initial PR was not meant to solve the footnotes behavior. It just gives us the tools to do so. I can include in the PR though.

**Comment by ellatrix** [comment]
@youknowriad Sorry, didn't meant to derail your PR :D

**Comment by youknowriad** [comment]
No worries @ellatrix in the last commit I went ahead and fixed footnotes undo/redo. There's one small remaining issue when we insert the first initial footnote, there's one undo level for the footnote addition and one for the block addition. That's because `onChange` is called twice between these two operations. Other than that, all is good in my testing.

**Comment by youknowriad** [comment]
@ellatrix I've just realized that you were already batching these two actions but it was still resulting in two onChange calls. The reason was a bug in nested `registry.batch` calls (we also have a batch call within RichText). This is fixed in the last commit (also added a unit test)

File: packages/core-data/src/entity-provider.js [hunk 3]
```diff
@@ -237,48 +236,57 @@ export function useEntityBlockEditor( kind, name, { id: _id } = {} ) {
...
 			// to make sure the edit makes the post dirty and creates
 			// a new undo level.
-			edits.content = ( { blocks: blocksForSerialization = [] } ) =>
-				__unstableSerializeAndClean( blocksForSerialization );
+			const edits = {
+				blocks: newBlocks,
+				selection,
+				content: ( { blocks: blocksForSerialization = [] } ) =>
+					__unstableSerializeAndClean( blocksForSerialization ),
+				...updateFootnotes( newBlocks ),
```
Comment (ellatrix) [inline packages/core-data/src/entity-provider.js:265]: Oh interesting, I didn't know there was a key for meta.

**Review by ellatrix — APPROVED** [review]
Undo works great for footnotes!

**Comment by tellthemachines** [comment]
I just cherry-picked this PR to the update/first-pre-beta-4 branch to get it included in the next release: ed76244b6f

**Comment by Mamaduka** [comment]
@youknowriad, I'm experimenting with this updated behavior/new flag. I wanted to check if the following behavior is expected or if I miss understanding the feature.

## Gist for the case
I've got three edits and I want to "cache" the second edit; the first two should be stored in the history stack.


### Expected undo stack
* three
* one
* original post title


### What I see now
* three
* two
* original post title


## Example code
```js
const edits = [ 'one', 'two', 'three' ];
const postId = wp.data.select( 'core/editor' ).getCurrentPostId();

edits.forEach(( edit, index ) => {
   // Cache the second edit.
   const isCached = index === 1;

    wp.data.dispatch( 'core' ).editEntityRecord( 'postType', 'post', postId, { title: edit }, { isCached } );
   
   console.log( { edit, isCached } );
});
```

**Comment by youknowriad** [comment]
I'm not entirely sure what you want to achieve but here's the behavior:

 1- When `isCached` is false: a new undo "step" is created for the change
 2- When `isCached` is true: no undo step is created for the change, instead the change is put into a "pending" area.
 3- Once a new edit is made that is supposed to create a new undo step (isCached: false) / or we explicitly call "createUndoStep",  the "pending" area is merged into the previous undo (So in this case, 1 is merged with 2 in the same undo step) and 3 creates a new undo step.

So we end up with `1+2    --->   3` as a history.

I'd say it's arguable whether we should merge 2 with 1 or 3 here but in my changes, I just left what we had for a long time because that was a proven behavior.

**Comment by Mamaduka** [comment]
This was just a simplified example. I was experimenting with the "coalescing edits" implementation in the `useEntityProp` hook, and the created undo stack wasn't matching my expectations.

Thanks for the explanation!
